### PR TITLE
Send Retry-After header with 429 response or 413 if samples exceed the burst of rate limiter

### PR DESCRIPTION
**Pull request by damnever**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by damnever** [comment]
`Retry-After` doesn't support duration less than one second, this is hard since we may overload our service or lose some tokens..

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -744,7 +744,27 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
...
+						rateLimitErr.Limit, validatedSamples, len(validatedMetadata), rateLimitErr.RetryAfter)),
+				})
+			}
+			return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
+				Code: http.StatusRequestEntityTooLarge,
+				Body: []byte(fmt.Sprintf("ingestion rate limit burst (%v) exceeded while adding %d samples and %d metadata, please break this request into multiple small requests",
+					rateLimitErr.Burst, validatedSamples, len(validatedMetadata))),
+			})
+		}
+		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples and %d metadata",
```
Comment (alvinlin123) [inline pkg/distributor/distributor.go:766]: Is there anyway we can avoid having the need of this line of code? Would be nice if we only have to deal with `limiter.RateLimitError` type.

**Comment by alvinlin123** [comment]
Hi damnever@ thank you for the PR, this is a nice change. May I ask you kindly add a CHANGELOG.md entry as well as add/update some tests? Thanks!

**Comment by damnever** [comment]
@alvinlin123 Somehow I have forgotten this patch. Please take another look.

**Comment by yeya24** [comment]
@damnever Thanks for the contribution!
I think this change makes sense. But for the case where the ingestion path is already overloaded then retrying could make it worse.
What about making it a flag on the distributor and we can choose the behavior of sending 429 retries or 413.

**Comment by damnever** [comment]
@yeya24 There are two changes:
1. add Retry-After header to the existing 429 responses.
2. change a part of 429 response to 413 since those requests contain too many samples, retry is useless until someone raises the limit(burst).

Maybe we should add an option to enable 413 since 413 may cause the client to drop samples without retrying.

**Comment by yeya24** [comment]
I see, thanks.
If we only send 429 when it is retryable and otherwise 413 then I think it is a reasonable approach.

**Review by yeya24 — APPROVED** [review]

**Pull request closed without merging**